api/query.go

```diff
@@ -1,130 +1,136 @@
-			msg := "Invalid direction '%s', valid options are: %s"
-			return 0, errors.New(fmt.Sprintf(msg, direction, options))
-	}
+type SliceQuery struct {
+	Vds       string  `form:"vds"       json:"vds"       binding:"required"`
+	Direction string  `form:"direction" json:"direction" binding:"required"`
+	Lineno    *int    `form:"lineno"    json:"lineno"    binding:"required"`
+	Sas       string  `form:"sas"       json:"sas"       binding:"required"`
 }
 
 type Endpoint struct {
 	StorageURL string
+	Protocol   string
 }
 
-func (e *Endpoint) Health(ctx *gin.Context) {
-	ctx.String(http.StatusOK, "I am up and running")
-}
-
-func (e *Endpoint) SliceMetadata(ctx *gin.Context) {
-	var query SliceQuery
-
-	if err := ctx.ShouldBind(&query); err != nil {
+func (e *Endpoint) sliceMetadata(ctx *gin.Context, query SliceQuery) {
+	conn, err := vds.MakeConnection(e.Protocol,  e.StorageURL, query.Vds, query.Sas)
+	if err != nil {
 		ctx.AbortWithError(http.StatusBadRequest, err)
 		return
 	}
 
-	querystr := ctx.Request.URL.Query()
-
-	delete(querystr, "vds")
-	delete(querystr, "direction")
-	delete(querystr, "lineno")
+	axis, err := vds.GetAxis(strings.ToLower(query.Direction))
+	if err != nil {
+		ctx.AbortWithError(http.StatusBadRequest, err)
+		return
+	}
 
-	url := fmt.Sprintf("azure://%v", query.Vds)
+	buffer, err := vds.SliceMetadata(*conn, *query.Lineno, axis)
+	if err != nil {
+		log.Println(err)
+		ctx.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
 
-	/*
-	 * This is super buggy as assumes that no other query-paramters are
-	 * present.
-	 */
-	cred := fmt.Sprintf(
-		"BlobEndpoint=%v;SharedAccessSignature=?%v",
-		e.StorageURL,
-		querystr.Encode(),
-	)
+	ctx.Data(http.StatusOK, "application/json", buffer)
+}
 
-	axis, err := getAxis(strings.ToLower(query.Direction))
+func (e *Endpoint) slice(ctx *gin.Context, query SliceQuery) {
+	conn, err := vds.MakeConnection(e.Protocol, e.StorageURL, query.Vds, query.Sas)
+	if err != nil {
+		ctx.AbortWithError(http.StatusBadRequest, err)
+		return
+	}
 
+	axis, err := vds.GetAxis(strings.ToLower(query.Direction))
 	if err != nil {
 		ctx.AbortWithError(http.StatusBadRequest, err)
 		return
 	}
 
-	buffer, err := vds.SliceMetadata(url, cred, *query.Lineno, axis)
+	buffer, err := vds.Slice(*conn, *query.Lineno, axis)
 	if err != nil {
 		log.Println(err)
 		ctx.AbortWithStatus(http.StatusInternalServerError)
 		return
 	}
 
-	ctx.Data(http.StatusOK, "application/json", buffer)
+	ctx.Data(http.StatusOK, "application/octet-stream", buffer)
 }
 
-func (e *Endpoint) Slice(ctx *gin.Context) {
+func sliceParseGetReq(ctx *gin.Context) (*SliceQuery, error) {
 	var query SliceQuery
 
-	if err := ctx.ShouldBind(&query); err != nil {
-		ctx.AbortWithError(http.StatusBadRequest, err)
-		return
+	q, err := url.QueryUnescape(ctx.Query("query"))
+	if err != nil {
+		return nil, err
 	}
 
-	querystr := ctx.Request.URL.Query()
+	err = json.Unmarshal([]byte(q), &query)
```

Comment:
Passing something bad, like `{ "direction" :  "inline" } `causes 500. Why 500?
```
2022/07/21 12:13:22 [Recovery] 2022/07/21 - 12:13:22 panic recovered:
GET /slice?query={%20%22direction%22%20:%20%20%22inline%22%20} HTTP/1.1

Reply:
Good catch. It's because I use json.Marshal for GET requests, which doesn't know about the binding tag in the SliceQuery struct. I.e. it's fine with missing fields and happily continues until we hit OpenVDS. which of course goes haywire on empty strings.

I cannot use gin.ShouldBind on Get request because it's urlencoded, so might need to manually verify that the fields are present